accepted/2023/dotnet-nuget-setapikey.md

```diff
@@ -0,0 +1,234 @@
+As the important message says on the docs page, on Windows it uses Windows' Data Protection API (DPAPI), which isn't supported on Linux or Mac, and [the .NET libraries team have no plans to support it since it will have different security promises than on Windows](https://github.com/dotnet/runtime/issues/22886).
+There is [an issue to consider another way of having less secure "encrypted" passwords](https://github.com/NuGet/Home/issues/1851) on Linux and Mac (like what Mono does for the `ProtectedData` API), but it hasn't been implemented yet.
+Therefore, in order for this feature not to be blocked, `<apikeys>` needs to support unencrypted (plain text) secrets (just as source credentials do).
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+- `dotnet nuget apikey set`
+
+```text
+Description:
+  Set the API key in a NuGet.config file.
+
+Usage:
+  dotnet nuget apikey set <VALUE> [options]
+
+Arguments:
+  <VALUE> The value of the API key. If neither --from-file or --environment-variable is specified, the value is treated as a string literal.
+
+Options:
+  -e, --environment-variable Treat the <VALUE> argument as the environment variable name to read, whose value will be used as the API key.
+  -f, --from-file            Treat the <VALUE> argument as a filename, whose contents contain the API key.
+  -s, --source <SOURCE>      The package source which this API should be used with. [Required]
+  --configfile               The NuGet.config file to write the API key to. [Default: c:\users\zivkan\AppData\Roaming\NuGet\NuGet.config]
+  --store-in-clear-text      Don't use .NET's ProtectedData APIs to encrypt API key. Required on platforms that don't support the API.
+  -?, -h, --help             Show command line help.
+```
+
+Note on `--configfile`:
+
+NuGet has been using `--configfile`, rather than `--config-file` for multiple years, so this "incorrect" name is used to remain consistent with other commands.
+
+Additionally, since API keys are always secrets, `--configfile` will always default to the user-profile `NuGet.config`. See ["which nuget.config file" in the rationale and drawbacks section](#which-nugetconfig-file).
+
+This allows the API to securely be added to `nuget.config`. An example for Azure Pipelines:
+
+```yaml
+steps:
+- script: dotnet nuget apikey set -e APIKEY
+  name: Set nuget.org API key
+  variables:
+    APIKEY: $(SECRET_API_KEY_FOR_NUGET_ORG)
+```
+
+Note on `--source`:
+
+The technical feasibility needs to be checked, but `NuGet.config` has a key `defaultPushSource`.
+When this is defined, `--source` should be optional and if `dotnet nuget apikey set` is run without specifying a value for `--source`, the default push source should be used.
```

Comment:
[NuGet.exe setapikey  command docs](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-setapikey#usage) state that `If <source> is omitted, nuget.org is assumed.`

Reply:
@donnie-msft I find the scenario a bit improbable. If SourceA is the default push source, why would the customer use `apikey set` without specifying a source, and expect it to be SourceB?  Kartheek did bring up an interesting point, that `nuget.exe setapikey` appears to default to nuget.org. So, SourceB being nuget.org is the only scenario that makes any sense to me. So with that knowledge, I can accept that security is more important than convenience and that we can make `--source` a required option.

@kartheekp-ms this spec is specifically about a new `dotnet nuget apikey set` command, not the `nuget.exe` command.  I don't know if the docs are correct, but I assume they are. 

@JonDouglas any feedback?  I don't think we need to (or should) carry over nuget.exe defaults in the dotnet CLI (msbuild have changed some defaults, for example). If my proposal goes accepted, then customers publishing packages to nuget.org, when they migrate from `nuget.exe setapikey` to `dotnet nuget apikey set` will need to start specifying nuget.org's service index URL (or source name, assuming they have it in their nuget.config file). While I think it would have been nice to use the default push source for `apikey set` and `apikey unset`, I think `nuget.exe`'s default behaviour adds risk, so perhaps we should make `--source` a mandatory option.

Reply:
>  I find the scenario a bit improbable. If SourceA is the default push source, why would the customer use apikey set without specifying a source, and expect it to be SourceB? 

@zivkan Are you saying humans never make mistakes?! 😲
What about scripts, can they not make assumptions and rely on a default, but be affected by a context change?

>  I can accept that security is more important than convenience and that we can make --source a required option.

Well, it sounds like you're slowly agreeing to do this anyway, but only because nuget.org is an implied default. Either way, I prefer consistency and security over hidden defaults. Open to hear arguments against it.

Reply:
I don't think we ought to carry over the defaults. This command as designed has a required field that will fail otherwise. I think that is reasonable design instead of assuming a central registry.